nexus/src/app/instance.rs

```diff
@@ -774,6 +774,108 @@ impl super::Nexus {
+    /// terminate its VMM process and unregister the instance.
+    pub(crate) async fn instance_force_terminate(
+        &self,
+        opctx: &OpContext,
+        instance_lookup: &lookup::Instance<'_>,
+    ) -> Result<InstanceAndActiveVmm, InstanceStateChangeError> {
+        let (.., authz_instance) =
+            instance_lookup.lookup_for(authz::Action::Modify).await?;
+
+        let state = self
+            .db_datastore
+            .instance_fetch_with_vmm(opctx, &authz_instance)
+            .await?;
+
+        // Is the instance currently incarnated in a VMM process?
+        let Some(vmm) = state.vmm() else {
+            // If the instance is already un-incarnated, we're all good.
+            debug!(
+                opctx.log,
+                "asked to force terminate an instance that has no VMM, \
+                 doing nothing";
+                "instance_id" => %authz_instance.id(),
+                "instance_state" => ?state.instance.runtime(),
+            );
+            debug_assert_eq!(
+                state.instance.runtime().nexus_state,
+                db::model::InstanceState::NoVmm,
+                "if an instance has no active VMM, it must be in the NoVmm \
+                 state (this is enforced by a DB check constraint)",
+            );
+            return Ok(state);
+        };
+
+        // Ladies and gentlemen, we got him!
+        let propolis_id = PropolisUuid::from_untyped_uuid(vmm.id);
+        let sled_id = SledUuid::from_untyped_uuid(vmm.sled_id);
+        let unregister_result =
+            self.instance_ensure_unregistered(&propolis_id, &sled_id).await;
+        match unregister_result {
+            // VMM unregistered, now process the state transition.
+            Ok(Some(state)) => {
+                info!(
+                    opctx.log,
+                    "instance terminated with extreme prejudice";
+                    "instance_id" => %authz_instance.id(),
+                    "vmm_id" => %propolis_id,
+                    "sled_id" => %sled_id,
+                );
+
+                self.notify_vmm_updated(opctx, propolis_id, &state).await?;
+            }
+            // If the returned state from sled-agent is `None`, then the instance
+            // was already unregistered. This may have been from a prior
+            // instance-ensure-unregistered call, but, since we observed an
+            // active VMM above, the current observed VMM generation doesn't
+            // know that the VMM is gone, so it is possible that the sled-agent
+            // has misplaced this instance. Therefore, we will attempt to mark
+            // the VMM as `Failed` at the generation after which we observed the
+            // VMM. This is safe to do here, because if the instance has been
+            // unregistered due to a race with another
+            // instance-ensure-unregistered request (rather than a sled-agent
```

Comment:
nit: this is also true if the VMM terminated normally, correct? (I *hope* the same logic holds; if it doesn't I think I've gravely misunderstood something!)

Reply:
Ah, yeah, that's correct.